app/src/main/res/raw/binarization_shader.glsl

```diff
@@ -3,11 +3,32 @@ precision mediump float;
 
 uniform vec2 resolution;
 uniform sampler2D framebuffer;
+uniform int windowSize;
+uniform float threshold;
 
 void main() {
-    vec2 uv = gl_FragCoord.xy / resolution;
+    float w = 1.0 / resolution.x;
+    float h = 1.0 / resolution.y;
+    vec2 d = vec2(w, h);
 
-    vec4 binarized = step(vec4(0.5), texture2D(framebuffer, uv));
+    vec2 uv = gl_FragCoord.xy / resolution + d/2.0;
 
-    gl_FragColor = binarized;
+
+    vec4 binarized = texture2D(framebuffer, uv);
+    float subImageSum = 0.0;
+
+    for (int i = 0; i < windowSize; i++)
+    {
+        for (int j = 0; j < windowSize; j++)
+        {
+            subImageSum += texture2D(framebuffer, uv + h*float(i) + w*float(j)).x;
+            subImageSum += texture2D(framebuffer, uv - h*float(i) - w*float(j)).x;
+            subImageSum += texture2D(framebuffer, uv + h*float(i) - w*float(j)).x;
+            subImageSum += texture2D(framebuffer, uv - h*float(i) + w*float(j)).x;
+        }
+    }
```

Comment:
the middle pixel is counted 4 times because when j and i = 0, then all four calls have the same UV